docs/src/developer_documentation/ir_representation.md

```diff
@@ -0,0 +1,560 @@
+Similarly, the `IDGotoIfNot` is a direct translation of `Core.Compiler.GotoIfNot`, with the `dest` field being an `ID` rather than an `Int`.
+
+Furthermore, note that the `goto if not` instruction at the end of basic block `#2` now (correctly) jumps to basic block `#5`, whereas before it jumped to block `#4`.
+That is, by virtue of the fact that the `ID` associated to each basic block remains unchanged in `BBCode`, all pre-existing control flow relationships have remained the same.
+Moreover, we did not have to write any book-keeping code to ensure that this update happened correctly.
+
+Now that we've created the new basic block, we modify block `#3` to fall-through to the new block if `%2` is even, and to jump straight back to block `#2` if not:
+```jldoctest my_factorial
+julia> bb = bb_ir_copy.blocks[3];
+
+julia> cond_id = ID();
+
+julia> target_id = bb_ir_copy.blocks[2].inst_ids[1];
+
+julia> insert!(bb, 4, cond_id, new_inst(Expr(:call, iseven, target_id)));
+
+julia> bb.insts[end] = new_inst(IDGotoIfNot(cond_id, block_2_id));
+
+julia> new_ir = CC.IRCode(bb_ir_copy)
+  1 ─      nothing::Nothing
+4 2 ┄ %2 = φ (#1 => 1, #3 => %8)::Int64
+  │   %3 = φ (#1 => 0, #3 => %7)::Int64
+  │   %4 = Base.slt_int(%3, _2)::Bool
+  └──      goto #5 if not %4
+2 3 ─ %6 = (Core.Intrinsics.mul_int)(%3, 2)::Int64
+5 │   %7 = Base.add_int(%6, 1)::Int64
+6 │   %8 = (Core.Intrinsics.add_int)(%2, %7)::Int64
+2 │   %9 = (iseven)(%2)::Any
+  └──      goto #2 if not %9
+  4 ─      (println)(%2)::Any
+  └──      goto #2
+8 5 ─      return %2
+```
+Observe that in order to tie the conditional to the goto-if-not, we simply ensure that the `ID` associated to the instruction which computes the conditional appears in the `IDGotoIfNot` instruction.
+
+### Run the new code
+
+As ever, we can construct a `Core.OpaqueClosure` using `IRCode` in order to produce something runnable:
+```jldoctest my_factorial
+julia> oc = Core.OpaqueClosure(new_ir; do_compile=true)
+(::Int64)::Int64->◌
+
+julia> oc(1000)
+2
+12
+58
+248
+1014
+2037
+```
+Exactly what `oc` is computing is neither here nor there.
+The point is that we've successfully inserted a new basic block into Julia's IR, and produced a callable from it.
+
+## Summary
+
+We have reviewed the two representations of Julia IR used in Mooncake.
+Where possible, we always use `IRCode` -- as discussed, forwards-mode AD exclusively uses `IRCode`.
+`BBCode` is basically only needed when undertaking transformations which involve changes to basic block structure -- the insertion of new basic blocks, and the modification of terminators in a way which changes the predecessors / successors of a given block being the primary sources of these kinds of changes.
+Reverse-mode AD makes extensive use of such transformations, so `BBCode` is currently important there.
+
+There are various efforts to augment `IRCode` with the capability to handle changes to basic block structure in a convenient way.
```

Comment:
Perhaps add a ref to these efforts if they exist publically? 

```suggestion
There are various efforts to augment `IRCode` with the capability to handle changes to basic block structure in a convenient way.
```

Reply:
Good idea.